src/compiler.rkt

```diff
@@ -276,356 +276,289 @@
-               (max
-                (- (max xlo-exp yhi-exp) outlo-exp)
-                (- (max xhi-exp ylo-exp) outhi-exp))
-               (if (equal? (bigfloat-signbit outlo) (bigfloat-signbit outhi)) ; slack part
-                   0
-                   (get-slack))))]
+     (define slack (if (> x-exp 2)                    ; if x-exp > 2 (actually 2.718),
+                       30                             ; then at least 1 additional bit is needed
+                       0)) 
       
-    [(equal? op ival-pow)
-     ; log[Гpow] = max[ log(y) , log(y) + log[log(x)] ]
-     ;                                    ^^^^^^^^^^^ less than 30
-     (define xlo-exp (log2-approx (ival-lo (first srcs))))
-     (define xhi-exp (log2-approx (ival-hi (first srcs))))
-     (define ylo-exp (log2-approx (ival-lo (second srcs))))
-     (define yhi-exp (log2-approx (ival-hi (second srcs))))
-
-     (max 0 (+ (max ylo-exp yhi-exp)
-               (if (> (max xlo-exp xhi-exp) 2) ; if x-exp > 2 (actually 2.718),
-                   30                          ;    then at least 1 additional bit is needed
-                   0)))]
-
-    [(equal? ival-exp op)
-     (max 0
-          (log2-approx (ival-lo (car srcs)))
-          (log2-approx (ival-hi (car srcs))))]
-
-    [(equal? ival-tan op)
+     (list (max 0 y-exp)                              ; exponent per x
+           (max 0 (+ y-exp slack)))]                  ; exponent per y
+
+    [(ival-exp)
+     ; log[Гexp] = log[x]
+     (define x (car srcs))
+     (define x-exp (ival-max-log2-approx x))
+     (list (max 0 x-exp))]
+
+    [(ival-tan)
      ; log[Гtan] = log[x] - log[sin(x)*cos(x)] <= log[x] + |log[tan(x)]| + 1
      ;                                                      ^^^^^^^^^^^
-     ;                                                 tan can be (-inf, +inf) or around to zero
+     ;                                                      possible uncertainty
+     (define x (first srcs))
      (define outlo (ival-lo output))
      (define outhi (ival-hi output))
-       
-     (define out-exp
-       (+ 1 (max (abs (log2-approx outlo))
-                 (abs (log2-approx outhi)))
-          (if (equal? (bigfloat-signbit outlo) (bigfloat-signbit outhi))
-              0                            ; both bounds are positive or negative
-              (get-slack))))               ; tan is (-inf, +inf) or around zero
-
-     (max 0
-          (+ (log2-approx (ival-lo (car srcs))) out-exp)
-          (+ (log2-approx (ival-hi (car srcs))) out-exp))]
+     
+     (define x-exp (ival-max-log2-approx x))
+     (define out-exp-abs (max (abs (log2-approx outlo))
+                              (abs (log2-approx outhi))))
+     
+     (define slack (if (equal? (mpfr-sign outlo) (mpfr-sign outhi))
+                       0
+                       (get-slack)))                  ; tan is (-inf, +inf) or around zero
```

Comment:
If `x` is a small interval around `0` don't we want a small condition number, not a big one? Is this adding slack where we don't need it?